### PR TITLE
Add dReal

### PR DESCRIPTION
**Pull request by soonho-tri**
- Add dReal as an external project (see .gitmodule)
 - Add `WITH_DREAL` cmake option (default: OFF)
 - Add a dummy `DrealSolver`
 - Add a simple test `dreal_solver_test.cc`
 - Checked clang-format and google_styleguides.
 - Tested on {Ubuntu, OSX} x {gcc,clang}. dReal doesn't support Windows.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/3221)

### CODE REVIEW

**Comment by jamiesnape** [comment]
Ping me to update the CI with the `WITH_DREAL` option once its name has been approved by a reviewer.

**Comment by david-german-tri** [comment]
Why not default-on?

:lgtm: for platform review

+@sammy-tri for feature review w/r/t use of MathematicalProgram

---

Review status: 0 of 9 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3221#-:-KPrjyXvnwv_0tRVgwu3:b-hmke64)*

**Comment by soonho-tri** [comment]
@david-german-tri, I want to switch to 'default-on' when I add real `DrealSolver` class and all the plumbings are done.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3221#-:-KPrkxweZtFzeQW-B9za:b-cmzvxx)*

**Comment by jamiesnape** [comment]
Typically we switch every single external on for most of the CI builds, so I would suggest switching it to default-on now unless there is a very strong reason not to.

---

Review status: 0 of 9 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3221#-:-KPrlZ2j1KmLL8cgyg3x:b-exbmvh)*

**Comment by jamiesnape** [comment]
```
configure: cannot find Boost headers version >= 0.0.0
```
dReal requires Boost?

**Comment by soonho-tri** [comment]
>> configure: cannot find Boost headers version >= 0.0.0

> dReal requires Boost?

No. 

It's just the case that one of dReal's external package is checking Boost. I passed `--without-boost` but it still generates the warning.

**Comment by soonho-tri** [comment]
In terms of dependencies, dReal requires `autoconf`, `automake`, `bison`, `cmake(>=2.8.12)`, `flex`, `libtool`, `make`, `pkg-config`. dReal itself is using cmake, but requires autotools to build external packages.

**Comment by jamiesnape** [comment]
@drake-jenkins-bot retest this please

**Comment by jamiesnape** [comment]
@drake-jenkins-bot retest this please

**Comment by david-german-tri** [comment]
Still :lgtm:.  +@jamiesnape to confirm that list of dependencies won't explode anything in CI.

Also, you have merge conflicts.

---

Review status: 0 of 9 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3221#-:-KPt4ggwfdGGPJ0_9hF-:b-8fkv7j)*

**Comment by jamiesnape** [comment]
CI has all those dependencies, so we should be good. :lgtm:

---

Review status: 0 of 9 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3221#-:-KPt5UkqgUY6XAQEot27:b-aqd9lp)*

**Comment by soonho-tri** [comment]
> Also, you have merge conflicts.

Just rebased my commits on the current `master`.

**Comment by sammy-tri** [comment]
Reviewed 8 of 9 files at r1, 1 of 1 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[drake/solvers/dreal_solver.cc, line 12 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3221#-KPt4ixj8YHUsjBGKTCc:-KPt4ixkJSItJtBSc2zf:b-5wovnf) ([raw file](https://github.com/robotlocomotion/drake/blob/9052610e7a3b3176adc560048b5624b441b581ad/drake/solvers/dreal_solver.cc#L12)):*
> ```cpp
> namespace solvers {
> bool DrealSolver::available() const {
>   throw std::runtime_error("not implemented yet");
> ```

BTW it might make more sense to return false here?  I suppose it probably doesn't matter since nothing is calling it yet...

---

*[drake/solvers/test/dreal_solver_test.cc, line 13 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3221#-KPt4u7m2WGNtoZV8nv5:-KPt4u7m2WGNtoZV8nv6:bds7b1d) ([raw file](https://github.com/robotlocomotion/drake/blob/9052610e7a3b3176adc560048b5624b441b581ad/drake/solvers/test/dreal_solver_test.cc#L13)):*
> ```cpp
> namespace solvers {
> namespace {
> GTEST_TEST(testDreal, example1) {
> ```

Is this test only temporary to make sure dreal has been included in the build properly (and going to be replaced with a test using the DrealSolver interface once it's populated?)  If that's the case, I think a TODO indicating that it's going away would be appropriate.  If that's not the case, I'm not sure why we've got a test in Drake which only tests if an external is working correctly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3221)*

**Comment by soonho-tri** [comment]
*[drake/solvers/test/dreal_solver_test.cc, line 13 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3221#-KPt4u7m2WGNtoZV8nv5:-KPt7VG3UQ14lABn9LbC:b-m146ov) ([raw file](https://github.com/robotlocomotion/drake/blob/9052610e7a3b3176adc560048b5624b441b581ad/drake/solvers/test/dreal_solver_test.cc#L13)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> Is this test only temporary to make sure dreal has been included in the build properly (and going to be replaced with a test using the DrealSolver interface once it's populated?)  If that's the case, I think a TODO indicating that it's going away would be appropriate.  If that's not the case, I'm not sure why we've got a test in Drake which only tests if an external is working correctly.
</details>
I agree. I'll add a TODO comment here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3221)*

**Comment by soonho-tri** [comment]
*[drake/solvers/dreal_solver.cc, line 12 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3221#-KPt4ixj8YHUsjBGKTCc:-KPt7_S2b91WReVdMIR_:b-m5yhzv) ([raw file](https://github.com/robotlocomotion/drake/blob/9052610e7a3b3176adc560048b5624b441b581ad/drake/solvers/dreal_solver.cc#L12)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> BTW it might make more sense to return false here?  I suppose it probably doesn't matter since nothing is calling it yet...
</details>
Right. I'll fix it quickly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3221)*

**Comment by soonho-tri** [comment]
*[drake/solvers/dreal_solver.cc, line 12 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3221#-KPt4ixj8YHUsjBGKTCc:-KPtFHwV8DRihfSCwFJ4:bfj9y1b) ([raw file](https://github.com/robotlocomotion/drake/blob/9052610e7a3b3176adc560048b5624b441b581ad/drake/solvers/dreal_solver.cc#L12)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary>
> Right. I'll fix it quickly.
</details>
Changed (+commented).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3221)*

**Comment by soonho-tri** [comment]
*[drake/solvers/test/dreal_solver_test.cc, line 13 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3221#-KPt4u7m2WGNtoZV8nv5:-KPtFMfG6iNYnrmORgCr:bs3mhdg) ([raw file](https://github.com/robotlocomotion/drake/blob/9052610e7a3b3176adc560048b5624b441b581ad/drake/solvers/test/dreal_solver_test.cc#L13)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary>
> I agree. I'll add a TODO comment here.
</details>
Comment added.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3221)*

**Comment by sammy-tri** [comment]
:lgtm: (pending CI)

---

Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3221#-:-KPtL175EYtt19V1RrA6:b-lj32db)*

**Comment by sammy-tri** [comment]
@drake-jenkins-bot retest this please

**Comment by soonho-tri** [comment]
@sammy-tri, I'm fixing the problems causing the build failures (with the helps from @jamiesnape).

**Comment by sammy-tri** [comment]
Thanks, I must have misunderstood the reason for the failure and thought it might have been intermittent.

**Comment by soonho-tri** [comment]
@drake-jenkins-bot mac-clang-experimental please

**Comment by jamiesnape** [comment]
@drake-jenkins-bot linux-gcc-experimental-matlab please

File: drake/solvers/test/CMakeLists.txt [hunk 1]
```diff
@@ -24,3 +24,26 @@ target_link_libraries(gurobi_solver_test drakeOptimization
...
 endif()
+
+if(dreal_FOUND)
+  add_executable(dreal_solver_test dreal_solver_test.cc)
+  target_link_libraries(dreal_solver_test drakeOptimization
+    ${GTEST_BOTH_LIBRARIES})
+  add_test(NAME dreal_solver_test COMMAND dreal_solver_test)
+endif()
+
+add_matlab_test(NAME solvers/test/gloptipoly_constrained_test COMMAND gloptipoly_constrained_test)
```
Comment (jamiesnape) [inline drake/solvers/test/CMakeLists.txt]: Delete all these MATLAB tests. They were moved elsewhere.

**Comment by soonho-tri** [comment]
*[drake/solvers/test/CMakeLists.txt, line 35 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3221#-KPqptZDD2n7vW5pIBlx-r5-35:-KQ1xcm5iHXrEYiSDc_h:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/2efb1abf8b83cda3376892e4df8021d322d84bcd/drake/solvers/test/CMakeLists.txt#L35)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary>
> Delete all these MATLAB tests. They were moved elsewhere.
> 
> 
> 
> (https://github.com/RobotLocomotion/drake/blob/master/drake/matlab/solvers/test/CMakeLists.txt)
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3221)*

**Comment by soonho-tri** [comment]
*[CMakeLists.txt, line 259 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3221#-KQ1xgHutWhRAXWswcbl:-KQ1xgHutWhRAXWswcbm:b-n4xxve) ([raw file](https://github.com/robotlocomotion/drake/blob/6653e64d4c7f018c93e1e000fe87b40fd6a7e66e/CMakeLists.txt#L259)):*
> ```CMake
>     -DWITH_BULLET:BOOL=${WITH_BULLET}
>     -DWITH_DIRECTOR:BOOL=${WITH_DIRECTOR}
>     -DWITH_DREAL:BOOL=${WITH_DREAL}
> ```

@jamiesnape, I also added this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3221)*

**Comment by jwnimmer-tri** [comment]
:lgtm:

---

Reviewed 3 of 9 files at r1, 1 of 2 files at r3, 2 of 3 files at r4, 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[CMakeLists.txt, line 297 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3221#-KQ2-LeXIrJR16XQzik7:-KQ2-LeYZhFmviLtcUiT:b22m7ar) ([raw file](https://github.com/robotlocomotion/drake/blob/6653e64d4c7f018c93e1e000fe87b40fd6a7e66e/CMakeLists.txt#L297)):*
> ```CMake
>     mosek
>     nlopt
>     dreal
> ```

BTW Please alpha-sort this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3221#-:-KQ2-f0dixH-KXAx85wp:bnfp4nl)*

**Comment by soonho-tri** [comment]
*[CMakeLists.txt, line 297 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3221#-KQ2-LeXIrJR16XQzik7:-KQ2180XDfBH7rV1ZK1J:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/6653e64d4c7f018c93e1e000fe87b40fd6a7e66e/CMakeLists.txt#L297)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Please alpha-sort this.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3221)*

**Comment by jwnimmer-tri** [comment]
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3221)*

**Comment by jamiesnape** [comment]
@drake-jenkins-bot retest this please

**Pull request merged**